services/121-service/src/registrations-update-jobs/registrations-update-jobs.spec.ts

```diff
@@ -0,0 +1,137 @@
+import { UserService } from '@121-service/src/user/user.service';
+
+class RegistrationsServiceMock {
+  validateInputAndUpdateRegistration = jest.fn();
+}
+
+class RegistrationsUpdateJobEmailsServiceMock {
+  sendUpdateJobEmail = jest.fn();
+}
+
+class UserServiceMock {
+  findById = jest.fn().mockResolvedValue({
+    id: 456,
+    username: 'owner@example.com',
+    displayName: 'Owner User',
+  });
+}
+
+describe('RegistrationsUpdateJobsService', () => {
+  let service: RegistrationsUpdateJobsService;
+  let registrationsService: RegistrationsServiceMock;
+  let registrationsUpdateJobEmailsService: RegistrationsUpdateJobEmailsServiceMock;
+
+  beforeEach(async () => {
+    registrationsService = new RegistrationsServiceMock();
+    registrationsUpdateJobEmailsService =
+      new RegistrationsUpdateJobEmailsServiceMock();
+
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [
+        RegistrationsUpdateJobsService,
+        { provide: RegistrationsService, useValue: registrationsService },
+        {
+          provide: RegistrationsUpdateJobEmailsService,
+          useValue: registrationsUpdateJobEmailsService,
+        },
+        { provide: UserService, useClass: UserServiceMock },
+      ],
+    }).compile();
+
+    service = module.get<RegistrationsUpdateJobsService>(
+      RegistrationsUpdateJobsService,
+    );
+  });
+
+  it('should process registrations update job successfully', async () => {
+    // Arrange
+    const job: RegistrationsUpdateJobDto = {
+      programId: 123,
+      reason: 'Updating registrations',
+      data: [
+        { referenceId: 'reg-1', customAttribute: 'value1' },
+        { referenceId: 'reg-2', customAttribute: 'value2' },
+      ],
+      request: { userId: 456 },
+    };
+
+    // Act
+    await service.processRegistrationsUpdateJob(job);
+
+    // Assert
+    expect(
+      registrationsService.validateInputAndUpdateRegistration,
+    ).toHaveBeenCalledTimes(2);
+    expect(
+      registrationsUpdateJobEmailsService.sendUpdateJobEmail,
+    ).not.toHaveBeenCalled();
+  });
+
+  it('should process registrations update job with some failures and send email', async () => {
+    // Arrange
+    const job: RegistrationsUpdateJobDto = {
+      programId: 123,
+      reason: 'Updating registrations',
+      data: [
+        { referenceId: 'reg-1', customAttribute: 'value1' },
+        { referenceId: 'reg-2', customAttribute: 'value2' },
+      ],
+      request: { userId: 456 },
+    };
+
+    registrationsService.validateInputAndUpdateRegistration.mockImplementation(
+      async ({ referenceId }) => {
+        if (referenceId === 'reg-2') {
+          throw new Error('Update failed, reason: "bad, input" for reg-2');
+        }
+      },
+    );
+
+    // Act
+    await service.processRegistrationsUpdateJob(job);
+
+    // Assert
+    expect(
+      registrationsService.validateInputAndUpdateRegistration,
+    ).toHaveBeenCalledTimes(2);
```

Comment:
So in unit test in general I try to only assert the return object. However in cases like this, it is not possible because this function does not really return an object but sends and email instead. 

In this case I would only assert the internal functions that perform a write action (such as a POST request or a DB save)

So in this case I would remove this assertion as it does not perform write actions

```
   expect(
      registrationsService.validateInputAndUpdateRegistration,
Comment view
    ).toHaveBeenCalledTimes(2);
```

The reason is that you do not want to test the internal working of a function too much to make refactoring easier.

Reply:
I would also remove this assertion 
```
   expect(
      registrationsService.validateInputAndUpdateRegistration,
Comment view
    ).toHaveBeenCalledTimes(2);
```


from `should process registrations update job successfully`